p4-16/psa/PSA.mdk

```diff
@@ -1497,6 +1499,65 @@ table.
 A PSA implementation must implement tables that have both a
 `psa_direct_counter` and `psa_direct_meter` property.
 
+A PSA implementation need not support both `psa_implementation` and
+`psa_idle_timeout` properties on the same table.
+
+### Table entry timeout notification {#sec-idle-timeout}
+
+PSA uses the `psa_idle_timeout` to enable a table implementation send
+notifications from the PSA device when a configurable time has passed since an
+entry was last matched. The property may take one of two values --
+`NO_TIMEOUT`, and `NOTIFY_CONTROL`. `NO_TIMEOUT` disables idle timeout support
+for the table and it is the default value when the property is not present.
+`PSA_NOTIFY_CONTROL` enables the notification. A PSA implementation will then
+generate an API for the control plane to set time-to-live (TTL) values for table
+entries and if at any time during its lifetime, the table entry is not "hit"
+(&ie; not selected by any packet lookup) for a lapse of time greater or equal
+to its TTL, the device should generate a notification to the control plane.
+The rate and mode of how the notifications are generated and delivered to the
+control plane are subject to configuration parameters specified by the control
+plane API.
+
+Example:
+```
+enum PSA_IdleTimeout_t {
+  NO_TIMEOUT,
+  NOTIFY_CONTROL
+}
+
+table t {
+  action a1 () { ... }
+  action a2 () { ... }
+  key = { hdr.f1: exact; }
+  actions = { a1; a2; }
+  default_action = a2;
+  psa_idle_timeout = PSA_IdleTimeout_t.NOTIFY_CONTROL;
+}
+```
+
+Restrictions on the TTL values and notifications:
+
+- It is likely that any hardware implementation will have a limited number of
+  bits to represent the values, and, since the values are programmed at
+  runtime, it is the responsibility of the runtime (P4Runtime or other
+  controller software) to guarantee that the TTL values can be represented in
+  the device. This can be done by scaling the values to the number of bits
+  available on the platform, ensuring that the range of values between
+  different entries are representable. A PSA implementation should only enable
+  the programming of such tables, and return an error if the device does not
+  support the idle timeout at all.
+
+- If no value is programmed for a table entry, even though the table has
+  enabled the idle timeout property, the entry will not generate a
+  notification.
+
+- PSA does not require a timeout value for a default action entry. The reason
+  for not making this mandatory in the specification is that the default action
+  may not have an actual entry in the table and it is may be hit on the miss
```

Comment:
"it is may be hit on the miss table misses"?

Reply:
Replaced. Sounds a lot better :)